File: frontend/src/services/authService.ts
import { storage } from '@utils/storage';
import type { OAuthExchangeResponse, User } from '@types/index';

const APP_NAME = 'investment_agent';
const API_URL = import.meta.env.VITE_API_URL || '';
const WEB_CALLBACK_REDIRECT_URI = `${API_URL}/auth/web-callback/${APP_NAME}`;

export const authService = {
  /**
   * Get Google OAuth authorization URL and redirect the browser to Google login.
   * The redirect_uri_override tells myBackend to use the web callback endpoint
   * instead of the mobile deep link callback.
   */
  async loginWithGoogle(): Promise<void> {
    const params = new URLSearchParams({
      redirect_uri_override: WEB_CALLBACK_REDIRECT_URI,
    });
    const response = await fetch(`${API_URL}/login/google/${APP_NAME}?${params}`);
    if (!response.ok) {
      throw new Error('Failed to get Google auth URL');
    }
    const data = await response.json();
    window.location.href = data.authorization_url;
  },

  /**
   * Exchange OAuth code + state for JWT tokens.
   * Called from the AuthCallback page after Google redirects back.
   */
  async handleOAuthCallback(code: string, state: string): Promise<User> {
    const params = new URLSearchParams({
      code,
      service: 'GOOGLE',
      state,
      redirect_uri_override: WEB_CALLBACK_REDIRECT_URI,
    });
    const response = await fetch(`${API_URL}/auth/exchange/${APP_NAME}?${params}`);
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Authentication failed' }));
      throw new Error(error.detail || 'Authentication failed');
    }
    const data: OAuthExchangeResponse = await response.json();

    storage.setTokens(data.access_token, data.refresh_token);
    storage.setUser(data.user);

    return data.user;
  },

  /**
   * Logout user and clear stored data
   */
  logout(): void {
    storage.clearAll();
  },

  /**
   * Check if user is authenticated
   */
  isAuthenticated(): boolean {
    return !!storage.getAccessToken();
  },

  /**
   * Get stored user data
   */
  getStoredUser(): User | null {
    const userData = storage.getUser();
    if (!userData) return null;

    try {
      return JSON.parse(userData) as User;
    } catch {
      return null;
    }
  },
};

File: mobile app/lib/services/auth_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_web_auth_2/flutter_web_auth_2.dart';
import '../core/constants/api_constants.dart';
import '../core/storage/secure_storage.dart';
import '../models/auth.dart';
import '../models/user.dart';

class AuthService {
  final SecureStorage _storage;

  // Uses its own Dio instance (no auth interceptors) for the OAuth flow
  AuthService(this._storage);

  Future<User> initiateGoogleAuth() async {
    // OAuth uses production backend: Google requires a valid HTTPS domain as redirect URI.
    // Investment API calls (investments, portfolio, etc.) use local baseUrl instead.
    final oauthDio = Dio(BaseOptions(baseUrl: ApiConstants.oauthBaseUrl));

    // 1. Get Google OAuth URL from backend
    final urlResponse = await oauthDio.get(ApiConstants.googleAuthUrl);
    final authUrl = urlResponse.data['authorization_url'] as String;

    // 2. Open browser and intercept deep link callback (investtrack://auth?code=...&state=...)
    final result = await FlutterWebAuth2.authenticate(
      url: authUrl,
      callbackUrlScheme: 'investtrack',
    );

    // 3. Extract code and state from callback URL
    final uri = Uri.parse(result);
    final code = uri.queryParameters['code'];
    if (code == null) {
      throw Exception('Authorization code not found in callback');
    }
    final state = uri.queryParameters['state'] ?? '';

    // 4. Exchange code for tokens + user (also via production backend)
    final Response exchangeResponse;
    try {
      exchangeResponse = await oauthDio.get(ApiConstants.exchangeCode(code, state));
    } on DioException catch (e) {
      debugPrint('=== AUTH EXCHANGE ERROR ===');
      debugPrint('Status: ${e.response?.statusCode}');
      debugPrint('Body: ${e.response?.data}');
      debugPrint('URL: ${e.requestOptions.uri}');
      rethrow;
    }
    final data = OAuthExchangeResponse.fromJson(
      exchangeResponse.data as Map<String, dynamic>,
    );

    // 5. Store tokens and user
    await _storage.setTokens(data.accessToken, data.refreshToken);
    await _storage.setUser(data.user.toJson());

    return data.user;
  }

  Future<void> logout() async {
    await _storage.clearAuthData();
  }

  Future<bool> isAuthenticated() async {
    final token = await _storage.getAccessToken();
    return token != null;
  }

  Future<User?> getStoredUser() async {
    final userData = await _storage.getUser();
    if (userData == null) return null;
    return User.fromJson(userData);
  }
}
